### PR TITLE
Bug 1161927 - System Update API

### PR DESCRIPTION
**Pull request by gasolin**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c8bbf75ec729a4a69fa3588f73eab0f12022a87b

File: apps/system/js/updatable.js [hunk 1]
```diff
@@ -144,15 +144,53 @@ function SystemUpdatable() {
   this.downloading = false;
   this.paused = false;
   this.showingApplyPrompt = false;
+  this._updateProviderPromise = null;
+  this._idleObserver = null;
 
+  window.addEventListener('request-update-check', this);
```
Comment (timdream) [inline apps/system/js/updatable.js]: Please use `Service.register()` and `Service.request()`

File: apps/settings/js/panels/about/system_update_manager.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    ALREADY_LATEST_VERSION: 4,
+    OFFLINE: 5,
+    ERROR: 6,
+    UNKNOWN: -1
+  };
+
+  const APP_UPDATE_KEY = 'apps.updateStatus';
+
+  var SystemUpdateManager = Module.create(function SystemUpdateManager() {
+    if (!window.navigator.updateManager) {
```
Comment (gasolin) [inline apps/settings/js/panels/about/system_update_manager.js]: if API not available, we can not access the panel

File: apps/settings/js/panels/about/system_update_manager.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    return this._readyPromise;
+  };
+
+  /**
+   * possible return values:
+   *
+   * - for system updates:
+   *   - no-updates
+   *   - already-latest-version
+   *   - check-complete
```
Comment (gasolin) [inline apps/settings/js/panels/about/system_update_manager.js]: missing `active-update` stat

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8841270a5c4777e58a09c47cde853ac39e5fdf90

**Comment by gasolin** [comment]
need unit test for `app_update_manager.js` & `system_update_manager.js`

File: apps/system/js/update_manager.js [hunk 4]
```diff
@@ -813,23 +797,6 @@
       this.downloadDialog.dataset.nowifi = !this._wifiAvailable();
     },
 
-    checkForUpdates: function su_checkForUpdates(shouldCheck) {
-      if (!shouldCheck) {
-        return;
-      }
-
-      this._dispatchEvent('force-update-check');
```
Comment (gasolin) [inline apps/system/js/update_manager.js]: will remove `_dispatchEvent` method since it only used here

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=66c6b8d6ca0b26ef8cbcc337b996b425fe137e00

File: apps/system/js/updatable.js [hunk 5]
```diff
@@ -144,15 +144,53 @@ function SystemUpdatable() {
...
+      provider.addEventListener('progress', this);
+      provider.addEventListener('error', this);
+    }
+  });
+};
+
+SystemUpdatable.prototype._getUpdateProvider = function() {
+  if (!this._updateProviderPromise) {
+    this._updateProviderPromise =
+    window.navigator.updateManager.getProviders().then((providerInfos) => {
```
Comment (gasolin) [inline apps/system/js/updatable.js]: should follow system_update_manage to 
1. getActiveProvider first, then
2. if not exist, choose first provider instead

**Comment by gasolin** [comment]
Need fix Gij tests as well

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=34b993f3b0c0cef27d6cc7512ea518c650f25188

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7f914b1533aeb2a76b47bda459776104086d1bea

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=efcc52b1aff3cc8758b34ab8b1f5ede216516564

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=30fb88fa0741e2cbf3e711d1e84833dbbc2daace

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5385650b7be30c1eaca660e028a241918341d78d

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=62b80f5794fa40bf9cc106225c82d7172af740ec

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f8847e9b8fd1e81b6373a7e4cec58f4ac73bef60

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=15861fda9cc836c9f385f8101a6a4489de346180

**Pull request closed without merging**